src/v/datalake/protobuf_to_arrow_converter.cc

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#include "datalake/protobuf_to_arrow_converter.h"
+
+#include "datalake/errors.h"
+#include "datalake/proto_to_arrow_struct.h"
+
+#include <arrow/type.h>
+
+#include <memory>
+#include <stdexcept>
+
+namespace datalake {
+
+proto_to_arrow_converter::proto_to_arrow_converter(const ss::sstring& schema) {
+    initialize_protobuf_schema(schema);
+    if (!initialize_struct_converter()) {
+        throw initialization_error("Could not initialize arrow arrays");
+    }
+}
+arrow_converter_status
+proto_to_arrow_converter::add_message(const ss::sstring& serialized_message) {
+    std::unique_ptr<google::protobuf::Message> message = parse_message(
+      serialized_message);
+    if (message == nullptr) {
+        return arrow_converter_status::parse_error;
+    }
+
+    if (!_struct_converter->add_top_level_message(message.get()).ok()) {
+        return arrow_converter_status::internal_error;
+    }
+    return arrow_converter_status::ok;
+}
+arrow_converter_status proto_to_arrow_converter::finish_batch() {
+    if (!_struct_converter->finish_batch().ok()) {
+        return arrow_converter_status::internal_error;
+    }
+    return arrow_converter_status::ok;
+}
+std::shared_ptr<arrow::Table> proto_to_arrow_converter::build_table() {
+    auto table_result = arrow::Table::FromChunkedStructArray(
+      _struct_converter->finish());
+    if (table_result.ok()) {
+        return table_result.ValueUnsafe();
+    } else {
+        return nullptr;
+    }
+}
+std::vector<std::shared_ptr<arrow::Field>>
+proto_to_arrow_converter::build_field_vec() {
+    return _struct_converter->get_field_vector();
+}
+std::shared_ptr<arrow::Schema> proto_to_arrow_converter::build_schema() {
+    return arrow::schema(_struct_converter->get_field_vector());
+}
+
+void proto_to_arrow_converter::initialize_protobuf_schema(
```

Comment:
nit: looks like this is only used in the constructor. Maybe just inline it?

Also, doing work in a constructor always makes me do a double take, as it often indicates that we should instead have a static `proto_to_arrow_converter::create()` method or a factory or something. It's less surprising IMO to throw there (or return an error code), than in a constructor